### PR TITLE
Discv5 worker

### PR DESCRIPTION
**Pull request by wemeetagain**
WIP, opening for initial review / discussion.

Deployed on `nogroup-hzax41-0` and the `beta` group

### CODE REVIEW

File: packages/beacon-node/src/network/discv5/index.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    await this.status.workerApi.close();
+    await Thread.terminate((this.status.workerApi as unknown) as Thread);
+  }
+
+  onDiscoveredStr(enrStr: Uint8Array): void {
+    try {
+      this.emit("discovered", ENR.decode(Buffer.from(enrStr)));
+    } catch (e) {
+      // TODO there is a bug in enr encoding(?) that causes many enrs to be incorrectly encoded (and thus incorrectly decoded)
+      // this.logger.error("Unable to decode enr", {enr: Buffer.from(enrStr).toString("hex")}, e as Error);
```
Comment (wemeetagain) [inline packages/beacon-node/src/network/discv5/index.ts:78]: There seems to be a bug in ENR encoding?
Uncommenting this results in a flood of error logs.

File: packages/beacon-node/src/network/metadata.ts [hunk 2]
```diff
@@ -35,7 +34,7 @@ export interface IMetadataModules {
  * https://github.com/ethereum/consensus-specs/blob/v1.1.10/specs/phase0/p2p-interface.md#metadata
  */
 export class MetadataController {
-  private enr?: ENR;
+  private setEnrValue?: (key: string, value: Uint8Array) => Promise<void>;
```
Comment (wemeetagain) [inline packages/beacon-node/src/network/metadata.ts:37]: Someone has to 'own' the ENR.
Current choice is for the discv5 worker to own the ENR and just pass a function here to update it.

The FileENR should also be removed and we just write the ENR to disk on close.

**Comment by dapplion** [comment]
Good starting point! observables seems sufficient to fullfil the need of events here. To safely merge this we need high observability of the metrics of the worker. This blog is great at how to collect basically all data points we need from the worker https://prog.world/monitoring-multithreaded-node-js-applications/

Regarding metrics the approach of just fetching the metrics str is good for now. Also I think we should move to the pattern of the discv5 library creating the metrics names itself, like we do in gossipsub now.

**Comment by wemeetagain** [comment]
Screenshot from 2023-01-09 12-36-33

Added some vm-level metrics and appended to the discv5 dashboard

**Comment by wemeetagain** [comment]
@dapplion @tuyennhv let me know what kind of other metrics you'd like to see

**Comment by wemeetagain** [comment]
> Also I think we should move to the pattern of the discv5 library creating the metrics names itself, like we do in gossipsub now.

https://github.com/ChainSafe/discv5/pull/233 :)

**Comment by twoeths** [comment]
> @dapplion @tuyennhv let me know what kind of other metrics you'd like to see

looks great 👍 , can't think of other metrics to add

File: packages/beacon-node/src/network/discv5/index.ts (hunk 1 above)
Comment (dapplion) [reply packages/beacon-node/src/network/discv5/index.ts:78]: Any updates here?

**Review by philknows — APPROVED** [review]

**Pull request merged**